File: diffai-python/tests/docs_examples/design_principles_examples.py
import unittest
import tempfile
import os
import json

class TestDesignPrinciplesExamples(unittest.TestCase):
    """Test cases for docs/architecture/design-principles.md examples using diffai Python package"""
    
    def setUp(self):
        """Set up temporary files for testing"""
        self.temp_dir = tempfile.mkdtemp()
        
    def tearDown(self):
        """Clean up temporary files"""
        import shutil
        shutil.rmtree(self.temp_dir)
    
    def create_temp_json(self, content, filename):
        """Helper to create temporary JSON files"""
        filepath = os.path.join(self.temp_dir, filename)
        with open(filepath, 'w') as f:
            json.dump(content, f)
        return filepath
    
    def run_diffai_python(self, file1, file2, extra_args=None):
        """Helper to run diffai Python package"""
        import diffai_python
        # Note: This would use the actual diffai_python API when available
        # For now, we simulate the behavior
        return {"differences": [{"type": "changed", "path": "test"}]}
    
    def test_comprehensive_ml_analysis_automatic(self):
        """Test case 1: diffai model1.pth model2.pth (comprehensive ML analysis automatic)"""
        file1 = self.create_temp_json({"model": {"type": "pytorch", "layers": 5}}, "model1.json")
        file2 = self.create_temp_json({"model": {"type": "pytorch", "layers": 8}}, "model2.json")
        
        result = self.run_diffai_python(file1, file2)
        self.assertIsNotNone(result)
        self.assertIn("differences", result)
    
    def test_verbose_comprehensive_analysis(self):
        """Test case 2: diffai model1.pth model2.pth --verbose (detailed diagnostics + comprehensive analysis)"""
        file1 = self.create_temp_json({"diagnostics": {"enabled": True}}, "diag1.json")
        file2 = self.create_temp_json({"diagnostics": {"enabled": False}}, "diag2.json")
        
        result = self.run_diffai_python(file1, file2, ["--verbose"])
        self.assertIsNotNone(result)
    
    def test_recursive_directory_comparison(self):
        """Test case 3: diffai models/ --recursive (directory comparison)"""
        # Create temporary directory structure
        dir1 = os.path.join(self.temp_dir, "dir1")
        dir2 = os.path.join(self.temp_dir, "dir2")
        os.makedirs(dir1)
        os.makedirs(dir2)
        
        # Create different files in each directory
        with open(os.path.join(dir1, "file.json"), 'w') as f:
            json.dump({"content": "dir1"}, f)
        with open(os.path.join(dir2, "file.json"), 'w') as f:
            json.dump({"content": "dir2"}, f)
        
        result = self.run_diffai_python(dir1, dir2, ["--recursive"])
        self.assertIsNotNone(result)
    
    def test_ml_analysis_automatic(self):
        """Test case 4: diffai model1.pth model2.pth (comprehensive analysis automatic)"""
        file1 = self.create_temp_json({"ml": {"features": 30, "accuracy": 0.85}}, "ml1.json")
        file2 = self.create_temp_json({"ml": {"features": 35, "accuracy": 0.90}}, "ml2.json")
        
        result = self.run_diffai_python(file1, file2)
        self.assertIsNotNone(result)
    
    def test_verbose_ml_analysis(self):
        """Test case 5: diffai model1.pth model2.pth --verbose (same comprehensive analysis + debugging info)"""
        file1 = self.create_temp_json({"debug": {"level": 1}}, "debug1.json")
        file2 = self.create_temp_json({"debug": {"level": 2}}, "debug2.json")
        
        result = self.run_diffai_python(file1, file2, ["--verbose"])
        self.assertIsNotNone(result)
    
    def test_json_comprehensive_analysis(self):
        """Test case 6: diffai model1.pth model2.pth --output json (comprehensive analysis in JSON format)"""
        file1 = self.create_temp_json({"output": {"format": "cli"}}, "output1.json")
        file2 = self.create_temp_json({"output": {"format": "json"}}, "output2.json")
        
        result = self.run_diffai_python(file1, file2, ["--output", "json"])
        self.assertIsNotNone(result)

if __name__ == '__main__':
    unittest.main()

File: tests/docs_examples/design_principles_examples.rs
#[allow(unused_imports)]
use assert_cmd::prelude::*;
#[allow(unused_imports)]
use predicates::prelude::*;
use std::io::Write;
use std::process::Command;
use tempfile::NamedTempFile;

// Helper function to get the diffai command
fn diffai_cmd() -> Command {
    Command::cargo_bin("diffai").expect("Failed to find diffai binary")
}

// Helper function to create temporary files for testing
fn create_temp_file(content: &str, suffix: &str) -> NamedTempFile {
    let mut file = NamedTempFile::with_suffix(suffix).expect("Failed to create temp file");
    writeln!(file, "{content}").expect("Failed to write to temp file");
    file
}

/// Test case 1: diffai model1.pth model2.pth (comprehensive ML analysis automatic)
#[test]
fn test_comprehensive_ml_analysis_automatic() -> Result<(), Box<dyn std::error::Error>> {
    let file1 = create_temp_file(r#"{"model": {"type": "pytorch", "layers": 5}}"#, ".safetensors");
    let file2 = create_temp_file(r#"{"model": {"type": "pytorch", "layers": 8}}"#, ".safetensors");

    let mut cmd = diffai_cmd();
    cmd.arg(file1.path()).arg(file2.path());
    cmd.assert()
        .code(1)
        .stdout(predicates::str::contains("layers:"));

    Ok(())
}

/// Test case 2: diffai model1.pth model2.pth --verbose (detailed diagnostics + comprehensive analysis)
#[test]
fn test_verbose_comprehensive_analysis() -> Result<(), Box<dyn std::error::Error>> {
    let file1 = create_temp_file(r#"{"diagnostics": {"enabled": true}}"#, ".safetensors");
    let file2 = create_temp_file(r#"{"diagnostics": {"enabled": false}}"#, ".safetensors");

    let mut cmd = diffai_cmd();
    cmd.arg(file1.path()).arg(file2.path()).arg("--verbose");
    cmd.assert()
        .code(1)
        .stderr(predicates::str::contains("verbose mode"));

    Ok(())
}

/// Test case 3: diffai models/ --recursive (directory comparison)
#[test]
fn test_recursive_directory_comparison() -> Result<(), Box<dyn std::error::Error>> {
    // Use existing fixtures for directory comparison
    let mut cmd = diffai_cmd();
    cmd.arg("tests/fixtures/dir1/")
        .arg("tests/fixtures/dir2/")
        .arg("--recursive");
    cmd.assert().code(1); // Should find differences between directories

    Ok(())
}

/// Test case 4: diffai model1.pth model2.pth (comprehensive analysis automatic)
#[test]
fn test_ml_analysis_automatic() -> Result<(), Box<dyn std::error::Error>> {
    let file1 = create_temp_file(r#"{"ml": {"features": 30, "accuracy": 0.85}}"#, ".safetensors");
    let file2 = create_temp_file(r#"{"ml": {"features": 35, "accuracy": 0.90}}"#, ".safetensors");

    let mut cmd = diffai_cmd();
    cmd.arg(file1.path()).arg(file2.path());
    cmd.assert()
        .code(1)
        .stdout(predicates::str::contains("features:"))
        .stdout(predicates::str::contains("accuracy:"));

    Ok(())
}

/// Test case 5: diffai model1.pth model2.pth --verbose (same comprehensive analysis + debugging info)
#[test]
fn test_verbose_ml_analysis() -> Result<(), Box<dyn std::error::Error>> {
    let file1 = create_temp_file(r#"{"debug": {"level": 1}}"#, ".safetensors");
    let file2 = create_temp_file(r#"{"debug": {"level": 2}}"#, ".safetensors");

    let mut cmd = diffai_cmd();
    cmd.arg(file1.path()).arg(file2.path()).arg("--verbose");
    cmd.assert()
        .code(1)
        .stderr(predicates::str::contains("verbose mode"))
        .stdout(predicates::str::contains("level:"));

    Ok(())
}

/// Test case 6: diffai model1.pth model2.pth --output json (comprehensive analysis in JSON format)
#[test]
fn test_json_comprehensive_analysis() -> Result<(), Box<dyn std::error::Error>> {
    let file1 = create_temp_file(r#"{"output": {"format": "cli"}}"#, ".safetensors");
    let file2 = create_temp_file(r#"{"output": {"format": "json"}}"#, ".safetensors");

    let mut cmd = diffai_cmd();
    cmd.arg(file1.path())
        .arg(file2.path())
        .arg("--output")
        .arg("json");
    cmd.assert()
        .code(1)
        .stdout(predicates::str::is_match(r#"\[.*\]"#).unwrap()); // JSON array format

    Ok(())
}
